velox/exec/GroupingSet.cpp

```diff
@@ -412,4 +444,241 @@ const HashLookup& GroupingSet::hashLookup() const {
+      } catch (const std::exception& e) {
+        return false;
+      }
+      return true;
+    }
+    candidate = candidate->parent();
+  }
+  return false;
+}
+} // namespace
+
+void GroupingSet::ensureInputFits(const RowVectorPtr& input) {
+  // Spilling is considered if this is a final or single aggregation and
+  // spillPath is set.
+  if (isPartial_ || !spillPath_.has_value()) {
+    return;
+  }
+  auto numDistinct = table_->numDistinct();
+  if (!numDistinct) {
+    // Table is empty. Nothing to spill.
+    return;
+  }
+  auto tableIncrement = table_->hashTableSizeIncrease(input->size());
+
+  auto rows = table_->rows();
+
+  auto [freeRows, outOfLineFreeBytes] = rows->freeSpace();
+  auto outOfLineBytes =
+      rows->stringAllocator().retainedSize() - outOfLineFreeBytes;
+  int64_t flatBytes = input->estimateFlatSize();
+
+  if (!tableIncrement && freeRows > input->size() &&
+      (outOfLineBytes == 0 || outOfLineFreeBytes >= flatBytes)) {
+    // Enough free rows for input rows and enough variable length free
+    // space for the flat size of the whole vector. If outOfLineBytes
+    // is 0 there is no need for variable length space.
+    return;
+  }
+  // If there is variable length data we take the flat size of the
+  // input as a cap on the new variable length data needed.
+  auto increment =
+      rows->sizeIncrement(input->size(), outOfLineBytes ? flatBytes : 0) +
+      tableIncrement;
+  auto tracker = mappedMemory_->tracker();
+  VELOX_CHECK(tracker);
+  // There must be at least 2x the increment in reservation.
+  if (tracker->getAvailableReservation() > 2 * increment) {
+    return;
+  }
+
+  // Check if can increase reservation. The increment is the larger of
+  // twice the maximum increment from this input and 1/4 of the current
+  // reservation.
+  auto targetIncrement =
+      std::max<int64_t>(increment * 2, tracker->getCurrentUserBytes() / 4);
+  if (maybeReserve(targetIncrement, *tracker)) {
+    return;
+  }
+  auto outOfLineBytesPerRow = outOfLineBytes / numDistinct;
+  auto rowsToSpill =
+      targetIncrement / (rows->fixedRowSize() + outOfLineBytesPerRow);
+
+  spill(
+      numDistinct - rowsToSpill,
+      outOfLineBytes - (rowsToSpill * outOfLineBytesPerRow));
+}
+
+void GroupingSet::spill(int64_t targetRows, int64_t targetBytes) {
+  if (!spiller_) {
+    auto rows = table_->rows();
+    auto types = rows->keyTypes();
+    types.insert(
+        types.end(), intermediateTypes_.begin(), intermediateTypes_.end());
+    std::vector<std::string> names;
+    for (auto i = 0; i < types.size(); ++i) {
+      names.push_back(fmt::format("s{}", i));
+    }
+    auto fileSize = mappedMemory_->tracker()->getCurrentUserBytes() / 4;
+    spiller_ = std::make_unique<Spiller>(
+        *rows,
+        [&](folly::Range<char**> rows) { table_->erase(rows); },
+        ROW(std::move(names), std::move(types)),
+        // Spill up to 4 partitions based on bits 29 and 30 of the hash number.
```

Comment:
Would you add a comment explaining that the choice of bits 29 and 30 is arbitrary?